src/byteset/mod.rs

```diff
@@ -0,0 +1,142 @@
+use memchr::{memchr, memchr2, memchr3, memrchr, memrchr2, memrchr3};
+mod scalar;
+
+#[inline]
+fn build_table(byteset: &[u8]) -> [u8; 256] {
+    let mut table = [0u8; 256];
+    for &b in byteset {
+        table[b as usize] = 1;
+    }
+    table
+}
+
+#[inline]
+fn should_use_table(set_size: usize) -> bool {
+    // This probably should depend on `size_of::<usize>` as well as the size
+    // of `haystack`.
+    set_size >= 4
+}
+
+#[inline]
+pub(crate) fn find(haystack: &[u8], byteset: &[u8]) -> Option<usize> {
+    match byteset.len() {
+        0 => return None,
+        1 => memchr(byteset[0], haystack),
+        2 => memchr2(byteset[0], byteset[1], haystack),
+        3 => memchr3(byteset[0], byteset[1], byteset[2], haystack),
+        n => {
+            if should_use_table(n) {
+                let table = build_table(byteset);
+                scalar::forward_search_bytes(haystack, |b| {
+                    table[b as usize] != 0
+                })
+            } else {
+                scalar::forward_search_bytes(haystack, |b| {
+                    scalar::forward_search_bytes(byteset, |sb| b == sb)
+                        .is_some()
+                })
```

Comment:
I don't think this code path is ever taken, right? I realize maybe it would if `should_use_table` changes its predicate, but it might be simpler to just drop it (and `should_use_table`) altogether?

Reply:
Yeah, I kept it while testing but it seems like it's generally better to use the table unless we have an optimized version.